miniapps/autodiff/admfem.hpp

```diff
@@ -0,0 +1,587 @@
+                rr[jj].setGradient(1.0);
+                tape.evaluate();
+                for(int ii=0;ii<state_size;ii++){
+                    jac(jj,ii)=aduu[ii].getGradient();
+                }
+                tape.clearAdjoints();
+                rr[jj].setGradient(0.0);
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+};
+
+template<template<typename, typename, typename, int, int> class CTD
+         , int state_size=1, int param_size=0>
+class QFunctionAutoDiff
+{
+public:
+
+    /// Evaluates a function for arguments vparam and uu.
+    /// The evaluatin is based on the operator() in the
+    /// user provided functor CTD.
+    double QEval(const Vector &vparam, Vector &uu)
+    {
+        CTD<double, const Vector, Vector, state_size, param_size> tf;
+        return tf(vparam,uu);
+    }
+
+    /// Provides the same functionality as QGrad.
+    void QVectorFunc(const Vector &vparam, Vector &uu, Vector &rr)
+    {
+        QGrad(vparam,uu,rr);
+    }
+
+    /// Returns the first derivative of CTD(...) with
+    /// respect to the active arguments proved in vector uu.
+    /// The length of rr is the same as for uu.
+    void QGrad(const Vector &vparam, Vector &uu, Vector &rr)
+    {
+
+#ifdef MFEM_USE_ADFORWARD
+        // use forward mode
+        CTD<ad::ADFloatType, const Vector, ad::ADVectorType,
+                        state_size, param_size> tf;
+        rr.SetSize(state_size);
+        {
+            ad::ADVectorType aduu(state_size);
+            for(int i=0;i<state_size;i++){
+                aduu[i].setValue(uu[i]);
+                aduu[i].setGradient(0.0);
+            }
+
+            ad::ADFloatType rez;
+
+            for(int ii=0;ii<state_size;ii++){
+                aduu[ii].setGradient(1.0);
+                rez=tf(vparam,aduu);
+                rr[ii]=rez.getGradient();
+                aduu[ii].setGradient(0.0);
+            }
+        }
+
+#else
+        typedef codi::RealReverse ADFType;
+        typedef TADVector<ADFType> ADFVector;
+
+        CTD<ADFType, const Vector, ADFVector, state_size, param_size> tf;
+        {
+            ADFVector aduu(state_size);
+            ADFType rez;
+            for(int i=0;i<state_size;i++){
+                aduu[i]=uu[i];
+            }
+
+            ADFType::TapeType& tape =ADFType::getGlobalTape();
+            typename ADFType::TapeType::Position pos=tape.getPosition();
+
+            tape.setActive();
+            for(int ii=0;ii<state_size;ii++){ tape.registerInput(aduu[ii]); }
+
+            rez=tf(vparam,aduu);
+            tape.registerOutput(rez);
+            tape.setPassive();
+
+            rez.setGradient(1.0);
+            tape.evaluate();
+            for(int i=0;i<state_size;i++){
+                rr[i]=aduu[i].getGradient();
+            }
+            tape.reset(pos);
+        }
+#endif
+    }
+
+    /// Provides same functionality as QHessian.
+    void QJacobian(mfem::Vector &vparam, mfem::Vector &uu, mfem::DenseMatrix &jac)
```

Comment:
Why does this exist?

Reply:
The issue with the naming comes from two cases below:

First case) The user starts with a scalar function. The gradients will be the first derivative, and the Hessian will hold the second derivatives.

Second case) The user starts with a vector function, and then the Jacobian is computed on this vector function.

Now let say a user starts implementation with a scalar function representing some kind of energy. The gradients will add to the residual, and they will be computed by AD. Next, the Hessian will add to the tangent matrix, and it will be computed by AD again. The code will go through testing, and the user is happy with the implementation. At his point, he/she will want to improve performance. The most straightforward modification would be implementing the first derivative as a vector function and applying AD to compute its Jacobian. The main code will not require any modifications - by having name duplication of Jacobian/Hessian and QVectorFunc/QGrad. If that is not the case, the user will have to go and manually change the main code.